File: sdk/typescript/src/extension-describe.ts
import {
  SDK_MIN_COMPOZY_VERSION,
  SDK_NAME,
  type ExtensionDescribeProcess,
} from "./extension-contract.js";
import { normalizeHostMethodList, normalizeStringList } from "./extension-runtime.js";
import type {
  CmdPaletteConfig,
  DescribePayload,
  DescribeHookEvent,
  DescribeProfile,
  DescribeResourcePath,
  ExtensionDefinition,
  ExtensionCommandGroupSpec,
  ExtensionToolRuntimeDescriptor,
} from "./types.js";
import type { RegisteredTool } from "./extension-contract.js";

interface ExtensionDescribeInput {
  definition: ExtensionDefinition;
  tools: ExtensionToolRuntimeDescriptor[];
  commandGroups: ExtensionCommandGroupSpec[];
  watchSourceKinds: string[];
  sdkVersion: string;
}

export function defaultExtensionDescribeProcess(): ExtensionDescribeProcess {
  return {
    argv: process.argv,
    stdout: process.stdout,
    setExitCode: code => {
      process.exitCode = code;
    },
  };
}

export function runExtensionDescribeMode(
  describeProcess: ExtensionDescribeProcess,
  describe: () => DescribePayload
): boolean {
  if (describeProcess.argv.at(-1)?.trim() !== "__describe") {
    return false;
  }
  describeProcess.stdout.write(`${JSON.stringify(describe())}\n`);
  describeProcess.setExitCode(0);
  return true;
}

export function cloneExtensionToolDescriptors(
  tools: Iterable<RegisteredTool>
): ExtensionToolRuntimeDescriptor[] {
  return [...tools].map(tool => ({
    ...tool.descriptor,
    ...(tool.descriptor.input_schema === undefined
      ? {}
      : { input_schema: structuredClone(tool.descriptor.input_schema) }),
    ...(tool.descriptor.output_schema === undefined
      ? {}
      : { output_schema: structuredClone(tool.descriptor.output_schema) }),
    capabilities: [...(tool.descriptor.capabilities ?? [])],
    ...(tool.descriptor.command === undefined
      ? {}
      : { command: structuredClone(tool.descriptor.command) }),
  }));
}

export function cmdPaletteViewIDs(config: CmdPaletteConfig | undefined): string[] {
  return (config?.views ?? []).map(view => view.id.trim()).filter(id => id.length > 0);
}

export function buildExtensionDescribePayload(input: ExtensionDescribeInput): DescribePayload {
  const command = input.definition.subprocess?.command.trim() ?? "";
  if (!command) {
    throw new Error("subprocess command is required");
  }
  const viewIDs = cmdPaletteViewIDs(input.definition.resources?.cmd_palette);

  return {
    name: input.definition.name.trim(),
    version: input.definition.version.trim(),
    ...(input.definition.description?.trim()
      ? { description: input.definition.description.trim() }
      : {}),
    provides: normalizeStringList(input.definition.capabilities?.provides),
    permissions: normalizeHostMethodList(input.definition.permissions?.requires),
    requires_env: normalizeStringList(input.definition.requires_env),
    profiles: normalizeDescribeProfiles(input.definition.profiles),
    resources: {
      skills: normalizeDescribeResourcePaths(input.definition.resources?.skills),
      loops: normalizeDescribeResourcePaths(input.definition.resources?.loops),
      agents: normalizeDescribeResourcePaths(input.definition.resources?.agents),
      automation: normalizeDescribeResourcePaths(input.definition.resources?.automation),
      layouts: normalizeDescribeResourcePaths(input.definition.resources?.layouts),
      ...(input.definition.resources?.cmd_palette === undefined
        ? {}
        : { cmd_palette: structuredClone(input.definition.resources.cmd_palette) }),
    },
    subprocess: {
      command,
      args: [...(input.definition.subprocess?.args ?? [])],
      env: { ...input.definition.subprocess?.env },
    },
    ...(input.definition.network_participation === undefined
      ? {}
      : {
          network_participation: {
            required: input.definition.network_participation.required,
            mode: input.definition.network_participation.mode.trim().toLowerCase(),
            channel_scopes: normalizeStringList(
              input.definition.network_participation.channel_scopes
            ),
          },
        }),
    tools: [...input.tools].sort((left, right) => left.handler.localeCompare(right.handler)),
    command_groups: input.commandGroups.map(group => ({ ...group })),
    hook_events: normalizeDescribeHookEvents(input.definition.supported_hook_events),
    watch_source_kinds: input.watchSourceKinds,
    ...(viewIDs.length === 0 ? {} : { cmd_palette_views: viewIDs }),
    sdk: {
      name: SDK_NAME,
      version: input.sdkVersion,
      protocol_version: "1",
      min_compozy_version: SDK_MIN_COMPOZY_VERSION,
    },
  };
}

function normalizeDescribeResourcePaths(
  resources: DescribeResourcePath[] | undefined
): DescribeResourcePath[] {
  const unique = new Map<string, DescribeResourcePath>();
  for (const resource of resources ?? []) {
    const normalized = { path: resource.path.trim(), profile: resource.profile?.trim() };
    unique.set(`${normalized.path}\u0000${normalized.profile ?? ""}`, {
      path: normalized.path,
      ...(normalized.profile ? { profile: normalized.profile } : {}),
    });
  }
  return [...unique.values()].sort(
    (left, right) =>
      left.path.localeCompare(right.path) || (left.profile ?? "").localeCompare(right.profile ?? "")
  );
}

function normalizeDescribeHookEvents(events: DescribeHookEvent[] | undefined): DescribeHookEvent[] {
  const unique = new Map<string, DescribeHookEvent>();
  for (const item of events ?? []) {
    const event = item.event.trim() as DescribeHookEvent["event"];
    const profile = item.profile?.trim();
    unique.set(`${event}\u0000${profile ?? ""}`, { event, ...(profile ? { profile } : {}) });
  }
  return [...unique.values()].sort(
    (left, right) =>
      left.event.localeCompare(right.event) ||
      (left.profile ?? "").localeCompare(right.profile ?? "")
  );
}

function normalizeDescribeProfiles(profiles: DescribeProfile[] | undefined): DescribeProfile[] {
  return [...(profiles ?? [])]
    .map(profile => {
      const defaults = profile.defaults ?? {};
      return {
        name: profile.name.trim(),
        ...(profile.color?.trim() ? { color: profile.color.trim() } : {}),
        ...(profile.icon?.trim() ? { icon: profile.icon.trim() } : {}),
        ...(profile.emoji?.trim() ? { emoji: profile.emoji.trim() } : {}),
        defaults: {
          ...(defaults.agent?.trim() ? { agent: defaults.agent.trim() } : {}),
          ...(defaults.provider?.trim() ? { provider: defaults.provider.trim() } : {}),
          ...(defaults.sandbox?.trim() ? { sandbox: defaults.sandbox.trim() } : {}),
        },
        credentials: [...(profile.credentials ?? [])]
          .map(credential => ({
            provider: credential.provider.trim(),
            slot: credential.slot.trim(),
          }))
          .sort(
            (left, right) =>
              left.provider.localeCompare(right.provider) || left.slot.localeCompare(right.slot)
          ),
      };
    })
    .sort((left, right) => left.name.localeCompare(right.name));
}

File: sdk/go/extension_describe.go
package compozysdk

import (
	"encoding/json"
	"io"
	"maps"
	"slices"
	"strings"

	"github.com/compozy/compozy/sdk/go/contracts"
)

const describeArgument = "__describe"

// Describe returns the deterministic build-time contract assembled from SDK registrations.
func (e *Extension) Describe() (contracts.DescribePayload, error) {
	if e == nil {
		return contracts.DescribePayload{}, NewInternalError("extension is required")
	}

	e.mu.RLock()
	defer e.mu.RUnlock()
	if err := e.definition.validate(); err != nil {
		return contracts.DescribePayload{}, err
	}
	if strings.TrimSpace(e.definition.Subprocess.Command) == "" {
		return contracts.DescribePayload{}, NewInvalidParamsError("subprocess command is required", nil)
	}

	tools := make([]contracts.ExtensionToolRuntimeDescriptor, 0, len(e.toolHandlers))
	for _, registered := range e.toolHandlers {
		descriptor := registered.descriptor
		tools = append(tools, contracts.ExtensionToolRuntimeDescriptor{
			ID:                  contracts.ToolID(descriptor.ID),
			Handler:             descriptor.Handler,
			Description:         descriptor.Description,
			FriendlyVerb:        descriptor.FriendlyVerb,
			Preview:             descriptor.Preview,
			InputSchema:         cloneRawMessage(descriptor.InputSchema),
			OutputSchema:        cloneRawMessage(descriptor.OutputSchema),
			InputSchemaDigest:   descriptor.InputSchemaDigest,
			OutputSchemaDigest:  descriptor.OutputSchemaDigest,
			ReadOnly:            descriptor.ReadOnly,
			Risk:                contracts.RiskClass(descriptor.Risk),
			RequiresInteraction: descriptor.RequiresInteraction,
			Capabilities:        slices.Clone(descriptor.Capabilities),
			Command:             cloneSDKCommandSpec(descriptor.Command),
		})
	}
	slices.SortFunc(tools, func(left, right contracts.ExtensionToolRuntimeDescriptor) int {
		return strings.Compare(left.Handler, right.Handler)
	})

	permissions := make([]string, 0, len(e.definition.Permissions.Requires))
	for _, method := range e.definition.Permissions.Requires {
		permissions = append(permissions, string(method))
	}

	return contracts.DescribePayload{
		Name:        strings.TrimSpace(e.definition.Name),
		Version:     strings.TrimSpace(e.definition.Version),
		Description: strings.TrimSpace(e.definition.Description),
		Provides:    requiredStringList(e.definition.Capabilities.Provides),
		Permissions: requiredStringList(permissions),
		RequiresEnv: normalizeStrings(e.definition.RequiresEnv),
		Resources: contracts.DescribeResources{
			Skills:     normalizeStrings(e.definition.Resources.Skills),
			Loops:      normalizeStrings(e.definition.Resources.Loops),
			Agents:     normalizeStrings(e.definition.Resources.Agents),
			Automation: normalizeStrings(e.definition.Resources.Automation),
			Layouts:    normalizeStrings(e.definition.Resources.Layouts),
		},
		Subprocess: contracts.DescribeSubprocess{
			Command: strings.TrimSpace(e.definition.Subprocess.Command),
			Args:    slices.Clone(e.definition.Subprocess.Args),
			Env:     cloneStringMap(e.definition.Subprocess.Env),
		},
		NetworkParticipation: normalizeDescribeNetworkParticipation(e.definition.NetworkParticipation),
		Tools:                tools,
		CommandGroups:        e.commandGroupsLocked(),
		HookEvents:           normalizeStrings(e.definition.SupportedHookEvents),
		WatchSourceKinds:     e.watchSourceKindsLocked(),
		SDK: contracts.DescribeSDKInfo{
			Name:              SDKName,
			Version:           e.sdkVersion,
			ProtocolVersion:   ProtocolVersion,
			MinCompozyVersion: MinCompozyVersion,
		},
	}, nil
}

func normalizeDescribeNetworkParticipation(
	value *NetworkParticipationRequirement,
) *contracts.DescribeNetworkParticipation {
	if value == nil {
		return nil
	}
	return &contracts.DescribeNetworkParticipation{
		Required: value.Required, Mode: strings.TrimSpace(value.Mode),
		ChannelScopes: normalizeStrings(value.ChannelScopes),
	}
}

func (e *Extension) writeDescribe(output io.Writer) error {
	payload, err := e.Describe()
	if err != nil {
		return err
	}
	encoder := json.NewEncoder(output)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(payload); err != nil {
		return NewInternalError("encode describe payload: " + err.Error())
	}
	return nil
}

func describeModeRequested(args []string) bool {
	return len(args) > 1 && strings.TrimSpace(args[len(args)-1]) == describeArgument
}

func requiredStringList(values []string) []string {
	result := normalizeStrings(values)
	if result == nil {
		return []string{}
	}
	return result
}

func cloneStringMap(values map[string]string) map[string]string {
	if len(values) == 0 {
		return nil
	}
	return maps.Clone(values)
}
